externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLBackendService.scala

```diff
@@ -33,9 +38,40 @@ import org.apache.kyuubi.session.SessionManager
  *               that this backend service holds to run [[org.apache.kyuubi.operation.Operation]]s.
  */
 class SparkSQLBackendService(name: String, spark: SparkSession)
-  extends AbstractBackendService(name) {
+  extends AbstractBackendService(name) with Logging {
 
   def this(spark: SparkSession) = this(classOf[SparkSQLBackendService].getSimpleName, spark)
 
   override val sessionManager: SessionManager = new SparkSQLSessionManager(spark)
+
+  override def renewDelegationToken(
+      sessionHandle: SessionHandle,
+      authFactory: KyuubiAuthenticationFactory,
+      tokenStr: String): Unit =
+    synchronized {
+      // We hacked `TCLIService.Iface.RenewDelegationToken` to transfer Credentials from Kyuubi
+      // Server to Spark SQL engine
+      val newCreds = KyuubiHadoopUtils.decodeCredentials(tokenStr)
+      val currentUser = UserGroupInformation.getCurrentUser
+      val oldCreds = KyuubiHadoopUtils.getCredentialsInternal(currentUser)
+
+      val updateCreds = new Credentials()
+      KyuubiHadoopUtils.getTokenMap(newCreds).foreach { case (alias, newToken) =>
+        val oldToken = oldCreds.getToken(alias)
+        if (oldToken == null) {
+          updateCreds.addToken(alias, newToken)
+        } else {
+          if (KyuubiHadoopUtils.getTokenIssueDate(newToken) >
+              KyuubiHadoopUtils.getTokenIssueDate(oldToken)) {
+            updateCreds.addToken(alias, newToken)
+          } else {
+            warn(s"Ignore token with earlier issue date: $newToken")
+          }
+        }
+      }
+      SparkContextHelper.updateDelegationTokens(
```

Comment:
shall we update delegation token if:
`KyuubiHadoopUtils.getTokenIssueDate(newToken) <= KyuubiHadoopUtils.getTokenIssueDate(oldToken)` ?

Reply:
How about `updateCreds.numberOfTokens() > 1` ?